File: scala/gildedrose/src/main/scala/Inventory.scala
import quality.BasicItemQualityCalculator

case class Inventory(items: Seq[Item]) {

  def update_quality: Inventory = {
    Inventory(items.map(item => {
      Item(
        item.name,
        calculate_sell_in(item),
        calculate_quality(item)
      )
    }))
  }

  private def calculate_quality(item: Item): Int = {
    if (item.name == "Aged Brie") {
      item.quality + 1
    } else {
      val calculator = new BasicItemQualityCalculator
      calculator.calculate(item)
    }
  }

  private def calculate_sell_in(item: Item): Int = {
    if (item.sell_in == 0) {
      item.sell_in
    } else {
      item.sell_in - 1
    }
  }
}

File: ruby/gildedrose/lib/inventory.rb
class Inventory

  attr_reader :items
  
  def initialize(items = [])
    @items = items
  end

  def update_quality
    # If we are not allowed to store the item type in the item.
    # Then we probably have to get the type from the name here
    # This may be true as items may not be required to know about their selling rule
    items.each(&:update_quality)
  end

  def add_item(item)
    @items << item
  end
end
